### DISCUSSION TITLE
Why are some subscribers called listeners while in reality (according to the implemented interface) they are subscribers?

### DISCUSSION THREAD

**Discussion by artyuum** [symfony/symfony#47414] [Q&A] [answered]
Doesn't that contribute to adding unneeded confusion into Symfony's codebase?

https://github.com/symfony/symfony/blob/8084eb83a44a3639a95f0860456d737b7f2751dd/src/Symfony/Component/Security/Http/EventListener/IsGrantedAttributeListener.php#L30

I understand that in [this folder](https://github.com/symfony/symfony/tree/6.2/src/Symfony/Component/Security/Http/EventListener), we have a mix of subscribers and listeners and someday they had to decide between "EventListener" / "EventSubscriber" when creating the folder, but in that case, I would have created two separate folders:
- EventListener
- EventSubscriber

That way, we have each file properly named and no room for confusion.

Another question, maybe we don't need to have both (listeners and subscribers) in Symfony's core packages?

I know that this change might not be trivial and would be BC but the point of this discussion is mainly to know the reason behind it and if there is some plan in the near (or distant) future to improve that.

**Comment by wouterj** [comment, marked as answer]
Subscribers are listeners that can register themselves (instead of requiring to be registered by the outside world).

In fact, for performance optimalisation, subscribers and listeners are both registered as listener during container compilation when using the Syntony framework.

I see no reason to change naming or namespace here.

**Reply by artyuum** [reply]
> Subscribers are listeners that can register themselves (instead of requiring to be registered by the outside world).

This only works inside a Symfony application with autoconfigure enabled, not when the subscriber is part of Symfony's core packages, right?

I see that the IsGrantedListener is still being registered manually, as a subscriber, not as a listener. See:
https://github.com/symfony/symfony/blob/076206a4d975358f0efafccd61bf81d31662f0ad/src/Symfony/Bundle/SecurityBundle/Resources/config/security.php#L278-L283

Unless there is something I'm missing about container compilation and IsGrantedListener is actually being registered as listener.

**Reply by stof** [reply]
They are *defined* as subscribers in the config. But at runtime, the listeners are registered directly for performance reasons (see https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php for the logic registering listeners in the container building process).

**Reply by artyuum** [reply]
When contributing to Symfony's core packages, as subscribers are being treated as listeners (if I understood correctly), is it still revelant to create subscribers? Or should I always create listeners?

**Reply by stof** [reply]
The benefit of subscribers is that the registration metadata is co-located with the code.

**Reply by artyuum** [reply]
I understand better now. Thank you :)

**Comment by tgalopin** [comment]
Also, a subscriber is a class while a listener is a method. Subscribers (classes) only declare their method as listeners thanks to the interface. A subscriber "includes" one or several listeners.
